README.md
# Animated Image List for Flutter

[![Pub](https://img.shields.io/pub/v/animated_image_list.svg)](https://pub.dev/packages/animated_image_list)

Flutter Animated image list with parallax effect and image lightbox .

<!-- [![Animated Image list](https://yt-embed.herokuapp.com/embed?v=EJ7FQjdHYkA)](https://www.youtube.com/watch?v=EJ7FQjdHYkA "Animated Image lis") -->

![gif](https://github.com/youssefali424/animated_image_list/blob/master/mainGif.gif?raw=true)

[<img src="https://img.youtube.com/vi/MGTt9eqJioc/0.jpg" width="49%">](https://youtu.be/MGTt9eqJioc) [<img src="https://img.youtube.com/vi/gTqUJB75Scc/0.jpg" width="49%">](https://youtu.be/gTqUJB75Scc)

[<img src="https://img.youtube.com/vi/iLn3OgV07KY/0.jpg" width="49%">](https://youtu.be/iLn3OgV07KY) [<img src="https://img.youtube.com/vi/FLpzHNzCSQ8/0.jpg" width="49%">](https://youtu.be/FLpzHNzCSQ8)

## Getting Started

Add the package to your pubspec.yaml:

```yaml
animated_image_list: ^0.5.0
```

In your dart file, import the library:

```Dart
import 'package:animated_image_list/AnimatedImageList.dart';
```

Instead of using a `ListView` create a `AnimatedImageList` Widget:

```Dart
 AnimatedImageList(
               images: arr.map((e) => e.urls.small).toList(),
               builder: (context, index, progress) {
                 return Positioned.directional(
                     textDirection: TextDirection.ltr,
                     bottom: 15,
                     start: 25,
                     child: Opacity(
                       opacity: progress > 1 ? (2 - progress) : progress,
                       child: Text(
                         arr[index].user.username ?? 'Anonymous',
                         style: TextStyle(
                             color: Colors.white,
                             fontSize: 25,
                             fontWeight: FontWeight.w500),
                       ),
                     ));
               },
               scrollDirection: Axis.vertical,
               itemExtent: 100,
               maxExtent: 400,
             ),
```

### Parameters:

| Name              | Description                                                                                                                                                                                                                                          | Required | Default value     |
| ----------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- | ----------------- |
| `images`          | A list of images url to display in the list by default it accepts urls if custom image needed use provider paramter                                                                                                                                  | required | -                 |
| `provider`        | Function which maps an url or image string to an image provider                                                                                                                                                                                      | -        | -                 |
| `itemExtent`      | not selected item size required to calculate animations                                                                                                                                                                                              | required | 150               |
| `maxExtent`       | selected item size required to calculate animations                                                                                                                                                                                                  | required | 400               |
| `scrollDirection` | List scroll direction horizontal or vertical                                                                                                                                                                                                         | -        | Axis.vertical     |
| `builder`         | Builder function that returns a widget to display over the image / `progress` from 0...1 item selection progress from 1...2 item leaving view could be usefull if you want to animate something like text opacity above / `index` current item index | -        | -                 |
| `placeHolder`     | Optional function which returns default placeholder for lightbox and error widget if image fails to load                                                                                                                                             | -        | kTransparentImage |


CHANGELOG.md
## [0.5.1] - First Release on pub

* Fix bugs
* add main gif to readme

## [1.0.0] - null safety

* Migrate to sound null safety
* Migrate to flutter >= 2.0 
* Add an example project

example/lib/main.dart
import 'package:animated_image_list/AnimatedImageList.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyImage {
  final String url;
  final String name;
  const MyImage(this.url, this.name);
}

const arr = <MyImage>[
  MyImage(
      "https://st.depositphotos.com/1428083/2946/i/600/depositphotos_29460297-stock-photo-bird-cage.jpg",
      "photo-bird-cage"),
  MyImage(
      "https://images.ctfassets.net/hrltx12pl8hq/4plHDVeTkWuFMihxQnzBSb/aea2f06d675c3d710d095306e377382f/shutterstock_554314555_copy.jpg",
      "color"),
  MyImage(
      "https://assets.website-files.com/5f204aba8e0f187e7fb85a87/5f210a533185e7434d9efcab_hero%20img.jpg",
      "cab_hero"),
  MyImage("https://img-19.ccm2.net/8vUCl8TXZfwTt7zAOkBkuDRHiT8=/1240x/smart/b829396acc244fd484c5ddcdcb2b08f3/ccmcms-commentcamarche/20494859.jpg",
      "zoom"),
  MyImage(
      "https://images.unsplash.com/photo-1612151855475-877969f4a6cc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MXx8aGQlMjBpbWFnZXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&w=1000&q=80",
      "sunset plant"),
];

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: AnimatedImageList(
          images: arr.map((e) => e.url).toList(),
          builder: (context, index, progress) {
            return Positioned.directional(
                textDirection: TextDirection.ltr,
                bottom: 15,
                start: 25,
                child: Opacity(
                  opacity: progress > 1 ? (2 - progress) : progress,
                  child: Text(
                    arr[index].name,
                    style: TextStyle(
                        color: Colors.white,
                        fontSize: 25,
                        fontWeight: FontWeight.w500),
                  ),
                ));
          },
          scrollDirection: Axis.vertical,
          itemExtent: 150,
          maxExtent: 400,
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
